clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -142,9 +142,26 @@ public void transitionToFailed() {
         transitionTo(MemberState.FAILED);
     }
 
+    @Override
+    public void tryJoin() {
+        resetEpoch();
+        transitionTo(MemberState.UNJOINED);
+    }
+
+    @Override
+    public void leaveGroup() {
+        resetEpoch();
```

Comment:
I don't get the full idea here. What's the intended usage for this? If the intention is to use the transition as a trigger to send the last HB, then `resetEpoch` is incorrect, because we need epoch -1 (or -2 if static member). If the intention is just to call this after the last HB has been sent, then how are you seeing we'll trigger that last HB? (ex. from the unsubsbcribe)

Reply:
Maybe I leave it as TODO, because I'm not sure how leaveGroup would integrate as well.

Reply:
OK, I remember why I wrote this. I think the epoch should never be -1/-2, it is only used for sending a leave group request.  When we need/want to leave the group, then I think we would first send the heartbeat with negative epoch, then transition the state to not_in_group, it should also join with epoch = 0.

Reply:
instead of reset the Epoch, which seems confusing, I made the epcoh = OptionalInt.empty() so that when the member tries to send a commit, it won't send the epoch after leaving the group.